wallet/src/account/txo_cache/mod.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{BTreeMap, BTreeSet};
+
+use common::{
+    chain::{OutPoint, OutPointSourceId, TxOutput},
+    primitives::Idable,
+};
+use wallet_types::{
+    account_id::AccountBlockHeight, wallet_block::WalletBlock, AccountTxId, WalletTx,
+};
+
+pub struct TxoCache {
+    blocks: BTreeMap<AccountBlockHeight, WalletBlock>,
+    txs: BTreeMap<AccountTxId, WalletTx>,
+    consumed: BTreeSet<OutPoint>,
+}
+
+impl TxoCache {
+    pub fn empty() -> Self {
+        Self {
+            blocks: BTreeMap::new(),
+            txs: BTreeMap::new(),
+            consumed: BTreeSet::new(),
+        }
+    }
+
+    pub fn new(
+        blocks: BTreeMap<AccountBlockHeight, WalletBlock>,
+        txs: BTreeMap<AccountTxId, WalletTx>,
+    ) -> Self {
+        let mut cache = Self::empty();
+        for (block_height, block) in blocks {
+            cache.add_block(block_height, block);
+        }
+        for (tx_id, tx) in txs {
+            cache.add_tx(tx_id, tx);
+        }
+        cache
+    }
+
+    pub fn blocks(&self) -> &BTreeMap<AccountBlockHeight, WalletBlock> {
+        &self.blocks
+    }
+
+    pub fn txs(&self) -> &BTreeMap<AccountTxId, WalletTx> {
+        &self.txs
+    }
+
+    pub fn outpoints(&self) -> &BTreeSet<OutPoint> {
+        &self.consumed
+    }
+
+    pub fn add_block(&mut self, block_height: AccountBlockHeight, block: WalletBlock) {
+        for input in block.kernel_inputs().iter() {
+            self.consumed.insert(input.outpoint().clone());
+        }
+        self.blocks.insert(block_height, block);
+    }
```

Comment:
Should we also call `add_tx()` for every transaction within `block` here?

Edit: ah, looking down, we're adding kernel inputs. All good, ignore this comment.